crates/cluster/src/keyspace.rs

```diff
@@ -0,0 +1,198 @@
+
+/// [`Keyspace`] version.
+pub type Version = u64;
+
+impl Keyspace {
+    /// Creates a new [`Keyspace`].
+    pub fn new(
+        operators: HashSet<node_operator::Idx>,
+        replication_strategy: ReplicationStrategy,
+        version: u64,
+    ) -> Result<Keyspace, CreationError> {
+        if operators.len() < REPLICATION_FACTOR {
+            return Err(CreationError::TooFewOperators(operators.len()));
+        }
+
+        if operators.len() > MAX_OPERATORS {
+            return Err(CreationError::TooManyOperators(operators.len()));
+        }
+
+        Ok(Keyspace {
+            operators,
+            shards: (),
+            replication_strategy,
+            version,
+        })
+    }
+
+    pub(crate) async fn calculate<Shards>(self) -> Keyspace<Shards>
+    where
+        Self: sealed::Calculate<Shards>,
+    {
+        sealed::Calculate::<Shards>::calculate_shards(self).await
+    }
+}
+
+impl Keyspace<Shards> {
+    /// Returns the [`Shard`] that contains the specified key.
+    pub fn shard(&self, key: u64) -> Shard {
+        Shard {
+            replica_set: *self.shards.0.shard_replicas(ShardId::from_key(key)),
+        }
+    }
+}
+
+impl<S> Keyspace<S> {
+    /// Returns an [`Iterator`] over [`node_operator`]s of this [`Keyspace`].
+    pub fn operators(&self) -> impl Iterator<Item = node_operator::Idx> + '_ {
+        self.operators.iter().copied()
+    }
+
+    /// Returns [`ReplicationStrategy`] of this [`Keyspace`].
+    pub fn replication_strategy(&self) -> ReplicationStrategy {
+        self.replication_strategy
+    }
+
+    /// Returns [`Version`] of this [`Keyspace`].
+    pub fn version(&self) -> Version {
+        self.version
+    }
+
+    pub(super) fn contains_operator(&self, idx: node_operator::Idx) -> bool {
+        self.operators.contains(&idx)
+    }
+
+    pub(super) fn require_diff<T>(&self, other: &Keyspace<T>) -> Result<(), SameKeyspaceError> {
+        if self.operators == other.operators
+            && self.replication_strategy == other.replication_strategy
+        {
+            return Err(SameKeyspaceError);
+        }
+
+        Ok(())
+    }
+}
+
+impl Shard {
+    /// Returns [`ReplicaSet`] assigned to this [`Shard`].
+    pub fn replica_set(&self) -> ReplicaSet {
+        self.replica_set
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum CreationError {
+    #[error("Too few operators within a Keyspace: {_0} < {REPLICATION_FACTOR}")]
+    TooFewOperators(usize),
+
+    #[error("Too many operators within a Keyspace: {_0} > {MAX_OPERATORS}")]
+    TooManyOperators(usize),
+}
+
+#[derive(Debug, thiserror::Error)]
+#[error("The new Keyspace doesn't differ from the old one")]
+pub struct SameKeyspaceError;
+
+pub(crate) mod sealed {
+    use std::future::Future;
+
+    #[allow(unused_imports)]
+    use super::*;
+
+    /// Trait to make `Keyspace<()>` and `Keyspace<Shards>` polymorphic.
+    pub trait Calculate<Shards> {
```

Comment:
`Shards` sounds like a concrete type rather than a generic type argument. It makes more sense to have `S` here.